Update Model Card for Mamba

ParagEkbote opened the pull request:
# What does this PR do?

As described in the issue, this PR updates the model card for Mamba. Please let me know if any modifications are required and I will make the necessary changes.

Refs #36979

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).

## Who can review?
@stevhliu

ParagEkbote committed:
update model card.

stevhliu commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+
+## Notes
+
+- The current implementation leverages the original cuda kernels: the equivalent of flash attention for Mamba are hosted in the [`mamba-ssm`](https://github.com/state-spaces/mamba) and the [`causal_conv1d`](https://github.com/Dao-AILab/causal-conv1d) repositories. Make sure to install them if your hardware supports them!
```
Remove the `\` at the end of the code block.

```suggestion
- The current implementation uses the original CUDA kernels. The FlashAttention equivalent implementation is hosted in the [mamba-ssm](https://github.com/state-spaces/mamba) and [causal_conv1d](https://github.com/Dao-AILab/causal-conv1d) repositories. Make sure to install them if your hardware supports it!
- Mamba stacks `mixer` layers which are equivalent to `Attention` layers. You can find the main logic of Mamba in the `MambaMixer` class.
- The example below demonstrates how to fine-tune Mamba with [PEFT](https://huggingface.co/docs/peft).

   ```py
   from datasets import load_dataset
   from trl import SFTTrainer
   from peft import LoraConfig
   from transformers import AutoTokenizer, AutoModelForCausalLM, TrainingArguments
   
   model_id = "state-spaces/mamba-130m-hf"
   tokenizer = AutoTokenizer.from_pretrained(model_id)
   model = AutoModelForCausalLM.from_pretrained(model_id)
   dataset = load_dataset("Abirate/english_quotes", split="train")
   training_args = TrainingArguments(
       output_dir="./results",
       num_train_epochs=3,
       per_device_train_batch_size=4,
       logging_dir='./logs',
       logging_steps=10,
       learning_rate=2e-3
   )
   lora_config =  LoraConfig(
           r=8,
           target_modules=["x_proj", "embeddings", "in_proj", "out_proj"],
           task_type="CAUSAL_LM",
           bias="none"
   )
   trainer = SFTTrainer(
       model=model,
       processing_class=tokenizer,
      args=training_args,
       peft_config=lora_config,
       train_dataset=dataset,
       dataset_text_field="quote",
   )
   trainer.train()
   \```
```

ParagEkbote replied:
Updated as discussed.

stevhliu reviewed (commented):
Thanks!

Maybe let's also add a quantization example with https://huggingface.co/state-spaces/mamba-2.8b-hf.

```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig

quantization_config = TorchAoConfig("int4_weight_only", group_size=128)
tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-2.8b-hf")
model = AutoModelForCausalLM.from_pretrained("state-spaces/mamba-2.8b-hf", torch_dtype=torch.bfloat16, quantization_config=quantization_config, device_map="auto",)
input_ids = tokenizer("Plants create energy through a process known as", return_tensors="pt").to("cuda")

output = model.generate(**input_ids)
print(tokenizer.decode(output[0], skip_special_tokens=True))
```

ParagEkbote committed:
Apply suggestions from code review

Co-authored-by: Steven Liu <59462357+stevhliu@users.noreply.github.com>

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
````diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+<hfoption id="transformers-cli">
+
+```bash
+echo -e "Plants create energy through a process known as" | transformers-cli run --task text-generation --model state-spaces/mamba-130m-hf --device 0
````
Updated as discussed.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+  <div class="flex flex-wrap space-x-1">
+    <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">
+    <img alt="FlashAttention" src="https://img.shields.io/badge/%E2%9A%A1%EF%B8%8E%20FlashAttention-eae0c8?style=flat">
+    <img alt="SDPA" src="https://img.shields.io/badge/SDPA-DE3412?style=flat&logo=pytorch&logoColor=white">
```
Updated as discussed.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+# Mamba
 
-The Mamba model was proposed in [Mamba: Linear-Time Sequence Modeling with Selective State Spaces](https://arxiv.org/abs/2312.00752) by Albert Gu and Tri Dao.
+[Mamba](https://huggingface.co/papers/2312.00752) is a new `state space model` architecture that rivals the classic Transformers. The architecture stacks `mixer` layers, which are the equivalent of `Attention` layers. The core logic of `mamba` is held in the `MambaMixer` class.
```
Updated as discussed.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+You can find all the original Mamba checkpoints under the [state-spaces](https://huggingface.co/state-spaces) organization.
 
-The abstract from the paper is the following:
+# Usage
```
Updated as discussed.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+[Mamba](https://huggingface.co/papers/2312.00752) is a new `state space model` architecture that rivals the classic Transformers. The architecture stacks `mixer` layers, which are the equivalent of `Attention` layers. The core logic of `mamba` is held in the `MambaMixer` class.
 
-This model is a new paradigm architecture based on `state-space-models`. You can read more about the intuition behind these [here](https://srush.github.io/annotated-s4/).
+You can find all the original Mamba checkpoints under the [state-spaces](https://huggingface.co/state-spaces) organization.
```
Updated as discussed.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
```diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
-)
-trainer.train()
+</hfoption>
+<hfoption id="transformers-cli">
```
Updated as discussed.

ParagEkbote committed:
update quantization example.

ParagEkbote commented:
> Thanks!
> 
> Maybe let's also add a quantization example with https://huggingface.co/state-spaces/mamba-2.8b-hf.
> 
> ```python
> import torch
> from transformers import AutoModelForCausalLM, AutoTokenizer, TorchAoConfig
> 
> quantization_config = TorchAoConfig("int4_weight_only", group_size=128)
> tokenizer = AutoTokenizer.from_pretrained("state-spaces/mamba-2.8b-hf")
> model = AutoModelForCausalLM.from_pretrained("state-spaces/mamba-2.8b-hf", torch_dtype=torch.bfloat16, quantization_config=quantization_config, device_map="auto",)
> input_ids = tokenizer("Plants create energy through a process known as", return_tensors="pt").to("cuda")
> 
> output = model.generate(**input_ids)
> print(tokenizer.decode(output[0], skip_special_tokens=True))
> ```

I've added this example and updated it as per the new AOBaseConfig-based approach since the approach used above is deprecated as mentioned in the [docs](https://huggingface.co/docs/transformers/quantization/torchao).

Could you please review the changes?

cc: @stevhliu

ParagEkbote committed:
update example.

ParagEkbote commented on docs/source/en/model_doc/mamba.md:
````diff
@@ -14,85 +14,72 @@ rendered properly in your Markdown viewer.
+</hfoption>
+<hfoption id="AutoModel">
+
+```py
````
Updated as discussed.

ParagEkbote committed:
update

stevhliu reviewed (approved):
LGTM!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37863). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
